### ISSUE TITLE
sqlite3.OperationalError: database or disk is full while over 150 GB free on the drive

### ISSUE THREAD

**Issue by lpryszcz** [nanoporetech/megalodon#16]
Hi, I've encountered another error - megalodon complained about lack of space on the disk, however there is over 150 GB free on my SSD. 
I've noticed at times (especially toward the end of the run) megalodon uses way over 40-50GB of RAM, although most of the time it's fine with just 3GB of RAM. Could that be an issue? Are you creating any databases in memory? 

![image](https://user-images.githubusercontent.com/988152/70177736-391be600-16db-11ea-9345-3eace3b51d13.png)

I've noticed some indices are created in memory: 
`./megalodon/mods.py:87:                 mod_index_in_memory=True, uuid_index_in_memory=True):`

Could that be the problem? 
This seems to be done toward the end of the run and some of my few days runs crashed toward the end and unfortunately nothing was stored in the `per_read_modified_base_calls.db`...

I guess this is related to `megalodon/megalodon_helper.py` - maybe do that a parameters specified by the user? 64GB for mapping in-memory sqlite3 db is crazy amount of RAM. Also another question, this is 64GB per process or overall? 
```
# allow 64GB for memory mapped sqlite file access
MEMORY_MAP_LIMIT = 64000000000
```

Full stack below. 

```
[09:45:58] Using canonical alphabet ACGT and modified bases Y=6mA Z=5mC.
[09:45:58] Loading reference.
[09:46:00] Preparing workers to process reads.
[09:46:00] Processing reads.
3 most common unsuccessful read types:
    13.0% (  80465 reads) : No alignment
     0.0% (    224 reads) : Likely out of memory error.
     -----
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 619525/619525 [88:35:31<00:00,  1.94read/s, ksamp/s=187]
[02:21:32] Unsuccessful processing types:
    13.0% (  80465 reads) : No alignment
     0.0% (    224 reads) : Likely out of memory error.
[02:21:32] Waiting for mods database to complete indexing.
Process Process-5:
Traceback (most recent call last):
  File "/home/lpryszcz/src/miniconda3/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/lpryszcz/src/miniconda3/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/home/lpryszcz/src/miniconda3/lib/python3.7/site-packages/megalodon/mods.py", line 721, in _get_mods_queue
    mods_db.create_data_covering_index()
  File "/home/lpryszcz/src/miniconda3/lib/python3.7/site-packages/megalodon/mods.py", line 373, in create_data_covering_index
    self.cur.execute('CREATE INDEX data_cov_idx ON data(' +
sqlite3.OperationalError: database or disk is full
[02:22:22] Spawning process to sort mappings
[02:22:22] Spawning modified base aggregation processes.
[02:22:22] Aggregating 0 variants and 0 modified base sites over reads.
[02:22:22] Waiting for mappings sort
```

**Comment by marcus1487** [comment]
For the first set of questions concerning in-memory databases and RAM usage. The databases held in memory are not actually stored in sqlite, but simple python dictionaries. Sqlite does not enable holding specific tables in memory, so this is a bit of a hack around this. This is necessary as other wise a sqlite index must be maintained on these databases during writing and this gets very expensive for even reasonably small runs. Looking at the code, the read_ids dictionary probably could be removed, but this is almost certainly not causing the RAM spike you are seeing as these are filled continuously throughout a run.

For the question regarding the memory mapped limit, this actually has to do with memory mapping (linking memory slots directly to the processor for faster access). This should not increase the RAM by this amount, but instead sets the size of the database to map in this fashion. For more details concerning memory mapping in sqlite see [this help page](https://www.sqlite.org/mmap.html).

Finally to what I believe is the core issue here. This error is coming after the message that the mods database index is being created. This database essentially copies the largest table from the database, but sorted by the reference position for access at aggregation time. Without this index the aggregation step would take much longer. Thus what I believe is happening is that the entire size of this index is likely requested at one time and this may in fact be larger than the 150 GB available on this storage device. This indexing step is also quite likely why the RAM is spiking as there should not be any large memory consumption within the megalodon internals at this stage of processing.

It is odd that the per-read mods database does not contain any data though. With the default settings, the database should retain all inserted data with an application failure. Have you set the `--database-safety 0` flag by any chance? With the default setting even with this type of failure the results should be stored and a post-processing step could be made to add the index once the data were moved to a location with more storage.

**Comment by lpryszcz** [comment]
Hi @marcus1487 , thanks a lot for the detailed reply. The run I'm trying is rather small, 130k reads this I'm surprised it requires larger computing resources. This even happens when I ran it on SSD with over 600 GB of free space, thus I suspect some memory related issue. Or is it using TMPDIR? 
I didn't use `--database-safety 0` and every time I'm looking all the processed data. I'll try it next time though. 

```
[11:25:02] Using canonical alphabet ACGT and modified bases Y=6mA Z=5mC.
[11:25:02] Loading reference.
[11:25:05] Preparing workers to process reads.
[11:25:05] Processing reads.
3 most common unsuccessful read types:
    13.1% (  17117 reads) : No alignment
     0.0% (     50 reads) : Likely out of memory error.
     0.0% (      1 reads) : Unexpected error
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 130649/130649 [101:29:23<00:00,  2.80s/read, ksamp/s=58.3]
******************** WARNING: Unexpected errors occured. See full error stack traces for first (up to) 50 errors in "unexpected_megalodon_errors.5233.err" ********************
[16:54:28] Unsuccessful processing types:
    13.1% (  17117 reads) : No alignment
     0.0% (     50 reads) : Likely out of memory error.
     0.0% (      1 reads) : Unexpected error
[16:54:28] Waiting for mods database to complete indexing.
Process Process-5:
Traceback (most recent call last):
  File "/home/lpryszcz/src/miniconda3/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/lpryszcz/src/miniconda3/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/home/lpryszcz/src/miniconda3/lib/python3.7/site-packages/megalodon/mods.py", line 722, in _get_mods_queue
    mods_db.create_data_covering_index()
  File "/home/lpryszcz/src/miniconda3/lib/python3.7/site-packages/megalodon/mods.py", line 374, in create_data_covering_index
    self.cur.execute('CREATE INDEX data_cov_idx ON data(' +
sqlite3.OperationalError: database or disk is full
[16:56:39] Spawning process to sort mappings
[16:56:39] Spawning modified base aggregation processes.
[16:56:39] Aggregating 0 variants and 0 modified base sites over reads.
[16:56:39] Waiting for mappings sort
```

**Comment by marcus1487** [comment]
Hi @lpryszcz , megalodon should not generally use this much compute resource, so I appreciate the report and help in debugging this issue. Doing some googling it appears that this issue may have to do with how sqlite was compiled on a system. There are a number of flags that control temp database creation and the location including the use of TEMPDIR. Megalodon uses the python interface in a pretty vanilla fashion, so I haven't tested it thoroughly. I'm wondering if you might be using a system python install that may have some weird sqlite settings. If so could you try an installation from a new virtual environment?

For the `--database-safety` comment, I would suggest running with the `--database-safety 2` setting. This way the data committed up until this error should be saved and the indexing step could be assessed more directly. To be more specific you should be able to complete this type of run (with the error) and then get the same error as above with the python code below:
```
import sqlite3
fn='megalodon_results/per_read_modified_base_calls.db'
db = sqlite3.connect(fn)
cur = db.cursor()
cur.execute('CREATE INDEX data_cov_idx ON data(score_pos, score_mod, score_read, score)')
```
From here we might be able to identify the core issue.  Something to try first might be increasing the page size `cur.execute('PRAGMA max_page_count = 2147483646')`. This will allow the database on disk to be as large as possible (2 TB set here). See [SO post here](https://stackoverflow.com/questions/16685016/sqlite3-operationalerror-database-or-disk-is-full-on-lustre).

**Comment by marcus1487** [comment]
@lpryszcz have you had a chance to test whether the indexing step is indeed the root cause of this issue?

**Comment by marcus1487** [comment]
TL;DR Set `SQLITE_TMPDIR` environment variable before running Megalodon to avoid this issue.

I've recently run into this issue again. This appears to be an issue with the default temporary storage location sqlite uses. [Here is a good post](https://stackoverflow.com/questions/23249843/sqlite3-vacuum-database-or-disk-is-full/23251896#23251896) on the locations SQLite will look to store temporary tables. Setting the `SQLITE_TMPDIR` environment variable solved this issue for me (on unix system). 

If one runs into this issue after a completed run, the database can be rescued (in my case even with `--database-safety 0` set) using the following python snippet:
```
from megalodon import mods
mods_db = mods.ModsDb('megalodon_results/per_read_modified_base_calls.db', read_only=False) 
# set max page count for larger databases
mods_db.cur.execute('PRAGMA max_page_count = 2147483646').fetchone()
mods_db.create_data_covering_index()
mods_db.close()
```

**Comment by marcus1487** [comment]
Updates in release 2.1 should help alleviate this issue in the future, hopefully without the need for setting the `SQLITE_TMPDIR` environment variable, but please do post if this issue persists.

**Comment by marcus1487** [comment]
The 2.2 release should resolve this issue without the need to set the `SQLITE_TMPDIR`. Please post here if this issue is observed with Megalodon version >= 2.2.
